crates/runtime/src/parking_spot.rs

```diff
@@ -0,0 +1,485 @@
+//! on a queue keyed by some address.
+//! - *Unparking* refers to dequeuing a thread from a queue keyed by some address
+//! and resuming it.
+//!
+//!
+//! # Example
+//! ```rust
+//! use std::sync::atomic::{AtomicU64, Ordering};
+//! use std::sync::Arc;
+//! use std::thread;
+//! use wasmtime_runtime::parking_spot::ParkingSpot;
+//!
+//! let parking_spot: Arc<ParkingSpot> = Arc::new(ParkingSpot::default());
+//! let atomic: Arc<AtomicU64> = Arc::new(AtomicU64::new(0));
+//! let atomic_key = Arc::as_ptr(&atomic) as u64;
+//!
+//! let handle = {
+//!     let parking_spot = parking_spot.clone();
+//!     let atomic = atomic.clone();
+//!     thread::spawn(move || {
+//!         atomic.store(1, Ordering::Relaxed);
+//!         parking_spot.unpark_all(atomic_key);
+//!         parking_spot.park(atomic_key, || atomic.load(Ordering::SeqCst) == 1, None);
+//!         assert_eq!(atomic.load(Ordering::SeqCst), 2);
+//!     })
+//! };
+//!
+//! parking_spot.park(atomic_key, || atomic.load(Ordering::SeqCst) == 0, None);
+//! atomic.store(2, Ordering::Relaxed);
+//! parking_spot.unpark_all(atomic_key);
+//!
+//! handle.join().unwrap();
+//! ```
+#![deny(clippy::all)]
+#![deny(clippy::pedantic)]
+#![deny(missing_docs)]
+#![deny(unsafe_code)]
+
+use std::collections::BTreeMap;
+use std::sync::{Arc, Condvar, Mutex};
+use std::time::{Duration, Instant};
+
+/// Result of a park operation.
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+pub enum ParkResult {
+    /// Unparked by another thread.
+    Unparked,
+
+    /// The validation callback returned false.
+    Invalid,
+
+    /// The timeout expired.
+    TimedOut,
+}
+
+#[derive(Default, Debug)]
+struct Spot {
+    /// The number of threads parked on this spot.
+    num_parked: u32,
+
+    /// The number of threads that have been unparked but not yet woken up.
+    /// This is used to avoid spurious wakeups.
+    to_unpark: u32,
+
+    /// The [`Condvar`] used to notify parked threads.
+    cvar: Arc<Condvar>,
+}
+
+/// The thread global `ParkingSpot`.
+#[derive(Default, Debug)]
+pub struct ParkingSpot {
+    inner: Mutex<BTreeMap<u64, Spot>>,
+}
+
+impl ParkingSpot {
+    /// Park the current thread until it is unparked or a timeout is reached.
+    ///
+    /// The `key` is used to identify the parking spot. If another thread calls
+    /// `unpark_all` or `unpark` with the same key, the current thread will be unparked.
+    ///
+    /// The `validate` callback is called before parking.
+    /// If it returns `false`, the thread is not parked and `ParkResult::Invalid` is returned.
+    ///
+    /// The `timeout` argument specifies the maximum amount of time the thread will be parked.
+    #[inline]
+    pub fn park(
+        &self,
+        key: u64,
+        validate: impl FnOnce() -> bool,
+        timeout: impl Into<Option<Duration>>,
+    ) -> ParkResult {
+        self.park_inner(key, validate, timeout.into())
+    }
+
+    #[inline]
```

Comment:
I don't think there's any need to tag all functions in this file with `#[inline]`

Reply:
removed